cryptoutil/util.go

```diff
@@ -148,18 +160,59 @@ func TryParsePEMBlock(block *pem.Block) (interface{}, error) {
 		return key, err
 	}
 
+	// Handle encrypted PEM blocks when a password is available
+	if len(password) > 0 {
+		// Legacy PEM encryption (Proc-Type: 4,ENCRYPTED)
+		if pt, ok := block.Headers["Proc-Type"]; ok && strings.Contains(pt, "ENCRYPTED") {
+			//nolint:staticcheck // Support legacy PEM encryption for backward compatibility.
+			decryptedDER, err := x509.DecryptPEMBlock(block, password)
+			if err != nil {
+				return nil, fmt.Errorf("failed to decrypt legacy encrypted PEM: %w", err)
+			}
+
+			// Parse based on likely key formats
+			if k, err := x509.ParsePKCS1PrivateKey(decryptedDER); err == nil {
+				return k, nil
+			}
+			if k, err := x509.ParseECPrivateKey(decryptedDER); err == nil {
+				return k, nil
+			}
+			if k, err := x509.ParsePKCS8PrivateKey(decryptedDER); err == nil {
+				return k, nil
+			}
+		}
+	}
+
 	return nil, ErrUnsupportedPEM{block.Type}
 }
 
+// TryParseKeyFromReader parses a key/cert from reader without password
+// support. For encrypted keys, use TryParseKeyFromReaderWithPassword.
 func TryParseKeyFromReader(r io.Reader) (interface{}, error) {
+	return TryParseKeyFromReaderWithPassword(r, nil)
+}
+
+// TryParseKeyFromReaderWithPassword parses a key/cert from reader, optionally
+// decrypting if a passphrase is provided.
+func TryParseKeyFromReaderWithPassword(r io.Reader, password []byte) (interface{}, error) {
 	bytes, err := io.ReadAll(r)
 	if err != nil {
 		return nil, err
 	}
 
+	if len(password) > 0 {
```

Comment:
Very minor nit: if you change this to len(password) == 0 and move the else into here, you reduce this function's nesting.  To me, it also reads more naturally to process simple cases up front.